instrumentation/nats/nats-2.17/testing/src/main/java/io/opentelemetry/instrumentation/nats/v2_17/AbstractNatsInstrumentationRequestTest.java

```diff
@@ -0,0 +1,332 @@
+        .runWithSpan(
+            "parent",
+            () -> connection.requestWithTimeout("sub", new byte[] {0}, Duration.ofSeconds(1)));
+
+    // then
+    assertCancellationPublishSpan();
+    assertTraceparentHeader(subscription);
+  }
+
+  @Test
+  void testRequestTimeoutFutureHeadersBody() throws InterruptedException {
+    // when
+    testing()
+        .runWithSpan(
+            "parent",
+            () ->
+                connection.requestWithTimeout(
+                    "sub", new Headers(), new byte[] {0}, Duration.ofSeconds(1)));
+
+    // then
+    assertCancellationPublishSpan();
+    assertTraceparentHeader(subscription);
+  }
+
+  @Test
+  void testRequestTimeoutFutureMessage() throws InterruptedException {
+    // given
+    NatsMessage message = NatsMessage.builder().subject("sub").data("x").build();
+
+    // when
+    testing()
+        .runWithSpan("parent", () -> connection.requestWithTimeout(message, Duration.ofSeconds(1)));
+
+    // then
+    assertCancellationPublishSpan();
+    assertTraceparentHeader(subscription);
+  }
+
+  @Test
+  void testRequestTimeoutFutureMessageHeaders() throws InterruptedException {
+    // given
+    NatsMessage message =
+        NatsMessage.builder().subject("sub").headers(new Headers()).data("x").build();
+
+    // when
+    testing()
+        .runWithSpan("parent", () -> connection.requestWithTimeout(message, Duration.ofSeconds(1)));
+
+    // then
+    assertCancellationPublishSpan();
+    assertTraceparentHeader(subscription);
+  }
+
+  private void assertPublishReceiveSpansSameTrace() {
+    testing()
+        .waitAndAssertTraces(
+            trace -> {
+              List<Consumer<SpanDataAssert>> asserts =
+                  new ArrayList<>(
+                      asList(
+                          // publisher: parent + publish
+                          span -> span.hasName("parent").hasNoParent(),
+                          span ->
+                              span.hasName("sub publish")
+                                  .hasKind(SpanKind.PRODUCER)
+                                  .hasParent(trace.getSpan(0))
+                                  .hasAttributesSatisfyingExactly(
+                                      messagingAttributes("publish", "sub", clientId)),
+                          // subscriber: process + publish(response)
+                          span ->
+                              span.hasName("sub process")
+                                  .hasKind(SpanKind.CONSUMER)
+                                  .hasParent(trace.getSpan(1)),
+                          span ->
+                              span.has(
+                                      new Condition<>(
+                                          data ->
+                                              data.getName().startsWith("_INBOX.")
+                                                  && data.getName().endsWith(" publish"),
+                                          "Name condition"))
```

Comment:
https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#span-name requires that destination is used in span name only when it has a low cardinality. This means that we should not produce spans with names like `_INBOX.NWzFVSx8azePx6VU3KZu8S.NWzFVSx8azePx6VU3KZuE1 publish`. We should find a way to detect such auto generated names and mark the destination as temporary, then the span naming logic won't use the destination name.